commit d75d57c2c06934a65feabd206129b9d1a230a1da
Author: varunkasyap <varunkasyap@hotmail.com>
Date:   Sat Nov 22 11:52:36 2025 +0530

    Fixed #36767 -- Allowed max redirect URL length to be set on HttpResponseRedirect.

diff --git a/tests/httpwrappers/tests.py b/tests/httpwrappers/tests.py
index 3e8364e616..b990e9f816 100644
--- a/tests/httpwrappers/tests.py
+++ b/tests/httpwrappers/tests.py
@@ -1,4 +1,5 @@
 import copy
+import itertools
 import json
 import os
 import pickle
@@ -488,15 +489,24 @@ class HttpResponseTests(SimpleTestCase):
 
     def test_redirect_url_max_length(self):
         base_url = "https://example.com/"
-        for length in (
-            MAX_URL_REDIRECT_LENGTH - 1,
-            MAX_URL_REDIRECT_LENGTH,
+        for length, response_class in itertools.product(
+            (MAX_URL_REDIRECT_LENGTH - 1, MAX_URL_REDIRECT_LENGTH),
+            (HttpResponseRedirect, HttpResponsePermanentRedirect),
         ):
             long_url = base_url + "x" * (length - len(base_url))
-            with self.subTest(length=length):
-                response = HttpResponseRedirect(long_url)
+            with self.subTest(length=length, response_class=response_class):
+                response = response_class(long_url)
                 self.assertEqual(response.url, long_url)
-                response = HttpResponsePermanentRedirect(long_url)
+
+    def test_redirect_url_max_length_override_via_param(self):
+        base_url = "https://example.com/"
+        for (max_length, length), response_class in itertools.product(
+            ((None, MAX_URL_REDIRECT_LENGTH + 1), (100, 99), (100, 100)),
+            (HttpResponseRedirect, HttpResponsePermanentRedirect),
+        ):
+            long_url = base_url + "x" * (length - len(base_url))
+            with self.subTest(length=length, response_class=response_class):
+                response = response_class(long_url, max_length=max_length)
                 self.assertEqual(response.url, long_url)
 
     def test_unsafe_redirect(self):
@@ -506,11 +516,23 @@ class HttpResponseTests(SimpleTestCase):
             "file:///etc/passwd",
             "é" * (MAX_URL_REDIRECT_LENGTH + 1),
         ]
-        for url in bad_urls:
-            with self.assertRaises(DisallowedRedirect):
-                HttpResponseRedirect(url)
-            with self.assertRaises(DisallowedRedirect):
-                HttpResponsePermanentRedirect(url)
+        for url, response_class in itertools.product(
+            bad_urls, (HttpResponseRedirect, HttpResponsePermanentRedirect)
+        ):
+            with (
+                self.subTest(url=url, response_class=response_class),
+                self.assertRaises(DisallowedRedirect),
+            ):
+                response_class(url)
+
+    def test_unsafe_redirect_via_max_length(self):
+        url = "https://example.com/"
+        for response_class in (HttpResponseRedirect, HttpResponsePermanentRedirect):
+            with (
+                self.subTest(response_class=response_class),
+                self.assertRaises(DisallowedRedirect),
+            ):
+                response_class(url, max_length=len(url) - 1)
 
     def test_header_deletion(self):
         r = HttpResponse("hello")
diff --git a/tests/shortcuts/tests.py b/tests/shortcuts/tests.py
index b80b8f5951..35942f846f 100644
--- a/tests/shortcuts/tests.py
+++ b/tests/shortcuts/tests.py
@@ -1,7 +1,9 @@
+from django.core.exceptions import DisallowedRedirect
 from django.http.response import HttpResponseRedirectBase
 from django.shortcuts import redirect
 from django.test import SimpleTestCase, override_settings
 from django.test.utils import require_jinja2
+from django.utils.http import MAX_URL_REDIRECT_LENGTH
 
 
 @override_settings(ROOT_URLCONF="shortcuts.urls")
@@ -56,3 +58,19 @@ class RedirectTests(SimpleTestCase):
                 )
                 self.assertIsInstance(response, HttpResponseRedirectBase)
                 self.assertEqual(response.status_code, expected_status_code)
+
+    def test_redirect_max_length_default_raises(self):
+        long_url = "https://example.com/" + "x" * MAX_URL_REDIRECT_LENGTH
+        msg = f"Unsafe redirect exceeding {MAX_URL_REDIRECT_LENGTH} characters"
+        with self.assertRaisesMessage(DisallowedRedirect, msg):
+            redirect(long_url)
+
+    def test_redirect_max_length_override_passes(self):
+        long_url = "https://example.com/" + "x" * MAX_URL_REDIRECT_LENGTH
+        response = redirect(long_url, max_length=None)
+        self.assertEqual(response.url, long_url)
+
+    def test_redirect_custom_strict_limit_raises(self):
+        msg = "Unsafe redirect exceeding 5 characters"
+        with self.assertRaisesMessage(DisallowedRedirect, msg):
+            redirect("https://example.com/", max_length=5)
